### PR TITLE
Kernels update

### PR DESCRIPTION
**Pull request by mp-17**
# Merge https://github.com/pulp-platform/ara/pull/81 and https://github.com/pulp-platform/ara/pull/101 before this one.

## Add baseline Jacobi2d, Dropout, Convolution3D benchmark

The convolution is now defined by its data type and its dimensions. 
`fconv3d`, for example, processes double-precision floating-point data, using 3D filters with depth `ch` (channels): 
`(i*i*ch)  ∗ (f*f*ch) = (o*o)`

Even if `fconv3d` is parameterized on the number of channels and can also be used with `ch = 1` becoming a `fconv2d`, the code for `fconv2d` is kept since it is more optimized for that particular case.

`fconv3d`: `F = {7}`, optimized with an enhanced algorithm

`fconv2d`: `F = {3, 7}`. F == 3 is optimized, F == 7 is optimized with an enhanced algorithm

`iconv2d`: `F = {3, 5, 7}`. F == 3 is optimized, F == 7 is optimized with an enhanced algorithm. F == 5 is not optimized

We will support and optimize the other filter sizes in the future.

The roofline plots produced for the convolutions are produced with the following parameters:
`iconv2d` = `F = 3 `
`fconv2d` = `F = 3 `
`fconv3d` = `F = 7 `

### Fixed

- Generate `data.S` files before compiling the programs
- Clean intermediate app object files with `make clean`
- Add a `fence` before stopping the cycle counter, to let the last vector store complete

### Added

- Add `fconv3d` kernel, optimized for 7x7 filters
- Optimize `fconv2d` and `iconv2d` kernels for 3x3 filters
- Add convolutions to the benchmark app, and print the related roofline plots
- Add corner case test to `vslidedown` instruction

### Changed

- Update README with instructions on how to compile convolutions
- Refactor `benchmark` app
- Double the testbench memory size
- Update the `python-requirements` list

## Checklist

- [x] Automated tests pass
- [x] Changelog updated
- [x] Code style guideline is observed

Please check our [contributing guidelines](CONTRIBUTING.md) before opening a Pull Request.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -42,6 +49,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Cut a timing-critical path from Addrgen to Sequencer (1 cycle more to start an AXI transaction)
 - Cut a timing-critical path in the `VSTU`, relative to the calculation of the pointer to the `VRF` word received from the lanes
 - Create `ara_system` wrapper containing Ara, Ariane, and an AXI mux, instantiated from within Ara's SoC
+- Update `README` with instructions on how to compile convolutions
+- Make `VSLDU` instruction queue parametric
+- Increase from 2 to 4 the `VSLDU` instruction queue depth
+- Increase from 2 to 4 the `VSLDU` operand queue depth
```
[the remark this answers is not in the document]
Comment (mp-17) [reply CHANGELOG.md]: Completely removed the hardware modifications from this PR. The source of the stalls was better addressed by https://github.com/pulp-platform/ara/pull/81

**Comment by mp-17** [comment]
Completely removed the hardware modifications from this PR. The source of the stalls was better addressed by #81.

**Comment by mp-17** [comment]
Cleaned version of this PR -> https://github.com/pulp-platform/ara/pull/104
